File: MMDB/MoviePosterUtil.m
//
//  MoviePosterUtil.m
//  MMDB
//
//  Created by Rob Wyland on 8/16/15.
//  Copyright (c) 2015 Org Inc. All rights reserved.
//

#import "MoviePosterUtil.h"
#import "Movie.h"

#define kSIZE  @[@"w92", @"w154"]
#define kIMG_URL @"http://image.tmdb.org/t/p/"

/**
 *  Enum the possible values supported
 *  Currently just a thumbnail and normal size
 */
typedef enum {
    ThumbnailSize = 0,
    NormalSize
} PosterImageSize;

@implementation MoviePosterUtil

+ (void)loadThumbnailPosterForMovie:(Movie *)movie intoView:(UIImageView *)imageView {
    if ( movie.poster_path == nil || [movie.poster_path isEqual:[NSNull null]] ) {
        // Sometimes there is no poster, so use placeholder
        [imageView setImage:[UIImage imageNamed:@"placeholder-w92.jpg"]];
        return;
    }
    
    UIImage *image = nil;
    if ( (image = [self lookupImageInTmpWithSize:ThumbnailSize Filename:movie.poster_path]) ) {
        // Found a cached value, use it
        [imageView setImage:image];
        return;
    }
    // Start as the placeholder
    [imageView setImage:[UIImage imageNamed:@"placeholder-w92.jpg"]];
    
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
        NSURL *url = [self downloadUrlForImageWithSize:ThumbnailSize filename:movie.poster_path];
        NSData *imageData = [[NSData alloc] initWithContentsOfURL:url];
        if ( imageData ) {
            [self updateView:imageView withImage:[UIImage imageWithData:imageData]];
            [self saveInTmpWithFile:imageData withSize:ThumbnailSize filename:movie.poster_path];
        }
    });
}

+ (void)loadPosterForMovie:(Movie *)movie intoView:(UIImageView *)imageView {
    if ( movie.poster_path == nil || [movie.poster_path isEqual:[NSNull null]] ) {
        // Sometimes there is no poster, so use placeholder
        [imageView setImage:[UIImage imageNamed:@"placeholder-w154.jpg"]];
        return;
    }
    
    UIImage *image = nil;
    if ( (image = [self lookupImageInTmpWithSize:NormalSize Filename:movie.poster_path]) ) {
        // Found a cached value, use it
        [imageView setImage:image];
        return;
    } else if ( (image = [self lookupImageInTmpWithSize:ThumbnailSize Filename:movie.poster_path]) ) {
        // Start as the thumbnail
        [imageView setImage:image];
    } else {
        // Start as the placeholder
        [imageView setImage:[UIImage imageNamed:@"placeholder-w154.jpg"]];
    }
    
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
        NSURL *url = [self downloadUrlForImageWithSize:NormalSize filename:movie.poster_path];
        NSData *imageData = [[NSData alloc] initWithContentsOfURL:url];
        if ( imageData ) {
            [self updateView:imageView withImage:[UIImage imageWithData:imageData]];
            [self saveInTmpWithFile:imageData withSize:NormalSize filename:movie.poster_path];
        }
    });
}

#pragma mark - Private Helper

/**
 *  Construct the url to request the resource
 */
+ (NSURL *)downloadUrlForImageWithSize:(PosterImageSize)imageSize filename:(NSString *)filename {
    // Construct the URL
    NSURL *url = [[NSURL alloc] initWithString:kIMG_URL];
    NSURL *sizeUrl = [url URLByAppendingPathComponent:kSIZE[imageSize]];
    NSURL *file = [sizeUrl URLByAppendingPathComponent:filename];

    return file;
}

/**
 *  Always update the View on the main thread
 *  Do a simple fade animation, this could always be nicer if need be
 */
+ (void)updateView:(UIImageView *)imageView withImage:(UIImage *)image {
    dispatch_async(dispatch_get_main_queue(), ^{
        [imageView setAlpha:0.f];
        [UIView animateWithDuration:0.25f animations:^{
            [imageView setImage:image];
            [imageView setAlpha:1.f];
        } completion:nil];
    });
}

/**
 *  Save the file contents into the tmp directory
 */
+ (void)saveInTmpWithFile:(NSData *)data withSize:(PosterImageSize)imageSize filename:(NSString *)filename {
    // Construct the filepath and append the size type
    NSString *tmpPath = NSTemporaryDirectory();
    NSString *filePath = [tmpPath stringByAppendingPathComponent:filename];
    NSString *file = [filePath stringByAppendingString:kSIZE[imageSize]];
    
    // Save file to tmp
    [data writeToFile:file atomically:YES];
}

/**
 *  Lookup an image in the tmp directory, and return it if it exists
 */
+ (UIImage *)lookupImageInTmpWithSize:(PosterImageSize)imageSize Filename:(NSString *)filename {
    // Construct the filepath and append the size type
    NSString *tmpPath = NSTemporaryDirectory();
    NSString *filePath = [tmpPath stringByAppendingPathComponent:filename];
    NSString *file = [filePath stringByAppendingString:kSIZE[imageSize]];
    
    NSError *error = nil;
    NSData *imageData = [NSData dataWithContentsOfFile:file options:0 error:&error];
    if ( error == nil ) {
        // Image found, return it
        return [UIImage imageWithData:imageData];
    }

    // Nothing found
    return nil;
}

@end

File: MMDB Swift/MoviePosterUtil.swift
//
//  MoviePosterUtil.swift
//  MMDB Swift
//
//  Created by Rob W on 12/11/15.
//  Copyright © 2015 Org Inc. All rights reserved.
//

import UIKit

let kSIZE: Array<String> = ["w92", "w154"]
let kIMG_URL: String = "http://image.tmdb.org/t/p/"

enum PosterImageSize: Int {
  case Thumbnail = 0, Normal
}

class MoviePosterUtil {
  
  /**
   *  Lookup the poster resource in size w92 and load into the imageView.
   *  Cache the values in the tmp directory in the app sandbox.
   */
  static func loadThumbnailPosterForMovie(movie: Movie, intoView imageView: UIImageView) {
    if let _ = movie.poster_path {
      if let cached = self.lookupImageInTmpWithSize(.Thumbnail, filename: movie.poster_path!) {
        // Found a cached value, use it
        imageView.image = cached
        return
      }
      
      // Start as the placeholder
      imageView.image = UIImage(imageLiteral: "placeholder-w92.jpg")
      dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), {
        let url: NSURL = downloadUrlForImageWithSize(.Thumbnail, filename: movie.poster_path!)
        if let imageData = NSData(contentsOfURL: url), image = UIImage(data: imageData) {
          updateView(imageView, withImage: image)
          saveInTmpWithFile(imageData, withSize: .Thumbnail, filename: movie.poster_path!)
        }
      })
    } else {
      // Sometimes there is no poster, so use placeholder
      imageView.image = UIImage(imageLiteral: "placeholder-w92.jpg")
    }
  }
  
  /**
   *  Lookup the poster resource in size w154 and load into the view.
   *  Temporarily stretch and use the w92 thumbnail (if available) while downloading the w154.
   *  Cache the values in the tmp directory in the app sandbox.
   */
  static func loadPosterForMovie(movie: Movie, intoView imageView: UIImageView) {
    if let _ = movie.poster_path {
      if let cached = self.lookupImageInTmpWithSize(.Normal, filename: movie.poster_path!) {
        // Found a cached value, use it
        imageView.image = cached
        return
      } else if let cached = self.lookupImageInTmpWithSize(.Thumbnail, filename: movie.poster_path!) {
        // Start as the thumbnail
        imageView.image = cached
      } else {
        // Start as the placeholder
        imageView.image = UIImage(imageLiteral: "placeholder-w154.jpg")
      }
      
      dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), {
        let url: NSURL = downloadUrlForImageWithSize(.Normal, filename: movie.poster_path!)
        if let imageData = NSData(contentsOfURL: url), image = UIImage(data: imageData) {
          updateView(imageView, withImage: image)
          saveInTmpWithFile(imageData, withSize: .Normal, filename: movie.poster_path!)
        }
      })
    } else {
      // Sometimes there is no poster, so use placeholder
      imageView.image = UIImage(imageLiteral: "placeholder-w154.jpg")
    }
  }

  // MARK: - Private Helper

  /**
   *  Construct the url to request the resource
   */
  static private func downloadUrlForImageWithSize(imageSize: PosterImageSize, filename: String) -> NSURL {
    // Construct the URL
    let url = NSURL(string: kIMG_URL)!
    let sizeUrl = url.URLByAppendingPathComponent(kSIZE[imageSize.rawValue])
    let file = sizeUrl.URLByAppendingPathComponent(filename)
    return file;
  }
  
  /**
   *  Always update the View on the main thread
   *  Do a simple fade animation, this could always be nicer if need be
   */
  static private func updateView(imageView: UIImageView, withImage image: UIImage ) {
    dispatch_async(dispatch_get_main_queue(), {
      imageView.alpha = 0
      UIView.animateWithDuration(0.25, animations: {
        imageView.image = image
        imageView.alpha = 1
      }, completion: nil)
    })
  }
  
  /**
   *  Save the file contents into the tmp directory
   */
  static private func saveInTmpWithFile(data: NSData, withSize imageSize: PosterImageSize, filename: String ) {
    // Construct the filepath and append the size type
    let tmpPath: String = NSTemporaryDirectory()
    let filePath: String = tmpPath + "/" + filename
    let file: String = filePath + kSIZE[imageSize.rawValue]
  
    // Save file to tmp
    data.writeToFile(file, atomically: true)
  }
  
  /**
   *  Lookup an image in the tmp directory, and return it if it exists
   */
  static private func lookupImageInTmpWithSize(imageSize: PosterImageSize, filename: String) -> UIImage? {
    let tmpPath: String = NSTemporaryDirectory()
    let filePath: String = tmpPath + "/" + filename
    let file: String = filePath + kSIZE[imageSize.rawValue]
   
    do {
      let image = try NSData(contentsOfFile: file, options: .DataReadingMappedIfSafe)
      return UIImage(data: image)
    } catch {
      return nil
    }
  }
}